python/TestHarness/schedulers/Scheduler.py

```diff
@@ -76,13 +135,143 @@ def __init__(self, harness, params):
+                    self.finished_jobs.add(tester.specs['test_name'])
+                else:
+                    self.skipped_jobs.add(tester.specs['test_name'])
+            else:
+                # Allow derived schedulers to perform postprocessing on results
+                self.returnToTestHarness(job_index, output)
+
+            # Close the output file
+            f.close()
+
+            # Clear this job slot
+            self.jobs[job_index] = None
+            self.slots_in_use = self.slots_in_use - slots
+
+    ## Returns False if all pre-preqs have been satisfied
+    def unsatisfiedPrereqs(self, tester):
+        if tester.getPrereqs() != [] and len(set(tester.getPrereqs()) - self.finished_jobs):
+            if self.options.ignored_caveats is None:
+                return True
+            elif self.options.ignored_caveats != 'all':
+                caveat_list = [x.lower() for x in self.options.ignored_caveats.split()]
+                if 'prereq' not in caveat_list:
+                    return True
+        return False
+
+    ## Schedule job for running
+    def schedule(self, tester, checks, test_list=None):
+        should_run = self.canLaunch(tester, checks, test_list)
+
+        # check for deprecated tuple
+        if type(should_run) == type(()):
+            (should_run, reason) = should_run
+            if not should_run:
+                reason = 'deprected checkRunnableBase #8037'
+                tester.setStatus(reason, tester.bucket_skip)
+                self.notifySchedulersBase(tester, add_to_table=True)
+                self.jobSkipped(tester.getTestName())
+
+        # This job is allowed to run
+        if should_run:
+            command = tester.getCommand(self.options)
+            # This method spawns another process and allows this loop to continue looking for tests
+            # RunParallel will call self.testOutputAndFinish when the test has completed running
+            # This method will block when the maximum allowed parallel processes are running
+            self.run(tester, command)
+
+        # This job is skipped - notify the scheduler
+        else:
+            if not tester.isSilent(): # SILENT occurs when a user is using --re options
+                if (self.options.report_skipped and tester.isSkipped()) \
+                   or tester.isSkipped():
+                    self.notifySchedulersBase(tester, add_to_table=True)
+                elif tester.isDeleted() and self.options.extra_info:
+                    self.notifySchedulersBase(tester, add_to_table=True)
+            self.jobSkipped(tester.getTestName())
+
+    ## Find an empty slot and launch the job
+    def launchJob(self, tester, command, slot_check=True):
+        # Make sure we are complying with the requested load average
+        self.satisfyLoad()
+
+        # Wait for a job to finish if the jobs queue is full
+        while self.jobs.count(None) == 0 or self.slots_in_use >= self.job_slots:
+            self.spinwait()
```

Comment:
This still doesn't feel right. It seems like the queuing classes should be much more simple. We get a job, we launch it and return to the test harness. Why are we doing slot counts and spin waits? We aren't running anything in the queuing class.